File: frontend/src/api/transactionLab.ts
import { apiGet, apiPost } from "./client";

export interface ExecuteResult {
  columns: string[] | null;
  rows: unknown[][] | null;
  row_count: number;
}

export async function beginTransaction(): Promise<{ session_id: string }> {
  return apiPost<{ session_id: string }>("/api/transaction-lab/begin");
}

export async function executeInTransaction(sessionId: string, sql: string): Promise<ExecuteResult> {
  return apiPost<ExecuteResult>("/api/transaction-lab/execute", { session_id: sessionId, sql });
}

export async function peekWithinTransaction(sessionId: string): Promise<ExecuteResult> {
  return apiPost<ExecuteResult>("/api/transaction-lab/peek", { session_id: sessionId });
}

export async function peekCommittedState(): Promise<ExecuteResult> {
  return apiGet<ExecuteResult>("/api/transaction-lab/peek-committed");
}

export async function commitTransaction(sessionId: string): Promise<{ status: string }> {
  return apiPost<{ status: string }>("/api/transaction-lab/commit", { session_id: sessionId });
}

export async function rollbackTransaction(sessionId: string): Promise<{ status: string }> {
  return apiPost<{ status: string }>("/api/transaction-lab/rollback", { session_id: sessionId });
}

File: backend/app/api/transaction_lab.py
import asyncio
from dataclasses import asdict
from typing import Any

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel

from app.services import transaction_lab

router = APIRouter(prefix="/api/transaction-lab", tags=["transaction-lab"])


class BeginResponse(BaseModel):
    session_id: str


class ExecuteRequest(BaseModel):
    session_id: str
    sql: str


class SessionRequest(BaseModel):
    session_id: str


class ExecuteResultSchema(BaseModel):
    columns: list[str] | None
    rows: list[list[Any]] | None
    row_count: int


class StatusSchema(BaseModel):
    status: str


@router.post("/begin", response_model=BeginResponse)
async def begin() -> BeginResponse:
    session_id = await asyncio.to_thread(transaction_lab.begin)
    return BeginResponse(session_id=session_id)


@router.post("/execute", response_model=ExecuteResultSchema)
async def execute(payload: ExecuteRequest) -> ExecuteResultSchema:
    try:
        result = await asyncio.to_thread(transaction_lab.execute, payload.session_id, payload.sql)
    except transaction_lab.TransactionLabError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    return ExecuteResultSchema(**asdict(result))


@router.post("/peek", response_model=ExecuteResultSchema)
async def peek(payload: SessionRequest) -> ExecuteResultSchema:
    try:
        result = await asyncio.to_thread(transaction_lab.peek_within_session, payload.session_id)
    except transaction_lab.TransactionLabError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    return ExecuteResultSchema(**asdict(result))


@router.get("/peek-committed", response_model=ExecuteResultSchema)
async def peek_committed() -> ExecuteResultSchema:
    result = await asyncio.to_thread(transaction_lab.peek_committed_state)
    return ExecuteResultSchema(**asdict(result))


@router.post("/commit", response_model=StatusSchema)
async def commit(payload: SessionRequest) -> StatusSchema:
    try:
        await asyncio.to_thread(transaction_lab.commit, payload.session_id)
    except transaction_lab.TransactionLabError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    return StatusSchema(status="committed")


@router.post("/rollback", response_model=StatusSchema)
async def rollback(payload: SessionRequest) -> StatusSchema:
    try:
        await asyncio.to_thread(transaction_lab.rollback, payload.session_id)
    except transaction_lab.TransactionLabError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    return StatusSchema(status="rolled_back")
